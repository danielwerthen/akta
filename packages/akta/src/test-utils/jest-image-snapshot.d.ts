declare module 'jest-image-snapshot' {
  export function toMatchImageSnapshot(received: any): any;
}
