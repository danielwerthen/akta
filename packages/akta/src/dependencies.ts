import { BehaviorSubject, firstValueFrom, Subject, switchMap } from 'rxjs';
import { createDependency, DependencyMap } from './dependency-map';
import { AllElements, isEvent } from './element-ops';
import { lazy } from './lazy-function';
import MetaObject, { MethodMissing } from './meta-object';
import { createSyncContext } from './sync-context';
import { AktaNode, ElementProperties } from './types';

export const dependecyContext = createSyncContext<DependencyMap>();

export const continuationDependency = createDependency(lazy());
export type TeardownFunction = () => void;
export const teardownDependency = createDependency<null | TeardownFunction[]>(
  null
);

export const elementsDependency = createDependency(new AllElements());

export function useNext<T extends AktaNode, RETURN extends AktaNode, NEXT>(
  _input?: () => Generator<T, RETURN, NEXT> | AsyncGenerator<T, RETURN, NEXT>
): (input: NEXT) => Promise<void> {
  const ctx = dependecyContext.getContext();
  return (input: NEXT) => {
    return firstValueFrom(
      ctx(continuationDependency).pipe(switchMap(next => next(input)))
    ).then(() => void 0);
  };
}

export function useTeardown(fn: TeardownFunction) {
  const ctx = dependecyContext.getContext();
  ctx.provide(teardownDependency, fns => {
    if (!fns) {
      return [fn];
    }
    fns.push(fn);
    return fns;
  });
}

export class PropertiesBase extends MetaObject<
  | ((factory?: () => Subject<unknown>) => Subject<unknown>)
  | ((start?: unknown) => Subject<unknown>)
> {
  [MethodMissing](key: string) {
    if (isEvent(key)) {
      return (factory?: () => Subject<unknown>) =>
        factory ? factory() : new Subject();
    } else {
      return (initialValue?: unknown) => {
        if (initialValue === undefined) {
          return new Subject();
        } else {
          return new BehaviorSubject(initialValue);
        }
      };
    }
  }
}

const propertiesBase = new PropertiesBase();
export class ElementPropertiesBase extends MetaObject<PropertiesBase> {
  [MethodMissing](_key: string) {
    return propertiesBase;
  }
}

export const Props: ElementProperties = (new ElementPropertiesBase() as unknown) as ElementProperties;
