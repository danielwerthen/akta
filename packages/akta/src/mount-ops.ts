import {
  combineLatest,
  Observable,
  tap,
  isObservable,
  filter,
  switchMap,
  of,
  Subject,
} from 'rxjs';

type TempNode = null | string | TempNode[] | Observable<TempNode>;

export class Attacher {
  private live: boolean = false;
  private nodes: ChildNode[] = [];
  private sibling: () => ChildNode | null;
  private parent: () => ChildNode;
  private activator: Subject<void> | undefined;
  private activators: Subject<void>[] = [];
  constructor(
    sibling: () => ChildNode | null,
    parent: () => ChildNode,
    activator?: Subject<void>
  ) {
    this.sibling = sibling;
    this.parent = parent;
    this.activator = activator;
    activator?.subscribe(() => {
      this.activator = undefined;
    });
  }
  remove(idx: number) {
    if (this.live && this.nodes[idx]) {
      this.nodes[idx].remove();
    }
    delete this.nodes[idx];
  }
  attach(node: ChildNode, idx: number) {
    if (this.live && this.nodes[idx]) {
      this.nodes[idx].remove();
    }
    this.nodes[idx] = node;
    if (this.live) {
      const sib = this.nodes[idx - 1] ?? this.sibling();
      if (sib) {
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(node);
      } else {
        const p = this.parent();
        if (p.childNodes.length > 0) {
          p.firstChild?.before(node);
        } else {
          p.appendChild(node);
        }
      }
    }
  }
  private _activate() {
    this.live = true;
    for (let i = 0; i < this.nodes.length; i++) {
      const node = this.nodes[i];
      if (!node) {
        continue;
      }
      const sib = this.nodes[i - 1] ?? this.sibling();
      if (sib) {
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(node);
      } else {
        const p = this.parent();
        if (p.childNodes.length > 0) {
          p.firstChild?.before(node);
        } else {
          p.appendChild(node);
        }
      }
    }
    for (let i = 0; i < this.activators.length; i++) {
      this.activators[i].complete();
    }
  }
  activate() {
    if (this.activator) {
      this.activator.subscribe(() => this._activate());
    } else {
      this._activate();
    }
  }
  branch(idx: number): Attacher {
    const activator = new Subject<void>();
    this.activators.push(activator);
    return new Attacher(
      () => this.nodes[idx - 1],
      this.parent,
      this.live ? undefined : activator
    );
  }
  teardown() {
    let node: ChildNode | undefined;
    while ((node = this.nodes.pop())) {
      if (this.live) {
        node.remove();
      }
    }
  }
}

export function* attach(sibling: () => ChildNode) {
  let lineup: ChildNode[] = [];
  let active = false;
  let yieldData: null | (() => ChildNode) = null;
  while (true) {
    const action:
      | 'activate'
      | [ChildNode, number]
      | 'teardown'
      | number = yield yieldData;
    yieldData = null;
    if (action === 'activate') {
      if (active) {
        throw new Error('Activate can only be triggered once');
      }
      active = true;
      for (let i = 0; i < lineup.length; i++) {
        const sib = lineup[i - 1] ?? sibling();
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(lineup[i]);
      }
    } else if (Array.isArray(action)) {
      const [node, idx] = action;
      if (active && lineup[idx]) {
        lineup[idx].remove();
      }
      lineup[idx] = node;
      if (active) {
        const sib = lineup[idx - 1] ?? sibling();
        if (!sib.parentNode) {
          throw new Error('Activation out of order');
        }
        sib.after(node);
      }
    } else if (action === 'teardown') {
      for (let i = 0; i < lineup.length; i++) {
        lineup[i].remove();
      }
      break;
    } else if (typeof action === 'number') {
      const idx: number = action;
      yieldData = () => lineup[idx - 1] ?? sibling();
    }
  }
}
function onlyFirst(_value: unknown, idx: number) {
  return idx === 0;
}

export function attachChildren(
  attacher: Attacher,
  children: TempNode,
  index?: number
): Observable<unknown> | void {
  if (!children) {
    const node = document.createTextNode('');
    attacher.attach(node, index ?? 0);
  } else if (typeof children === 'string') {
    const node = document.createTextNode(children);
    attacher.attach(node, index ?? 0);
  } else if (Array.isArray(children)) {
    const branch =
      typeof index === 'number' ? attacher.branch(index) : attacher;
    const observables = children
      .map((child, idx) => {
        return attachChildren(branch, child, idx);
      })
      .filter<Observable<unknown>>(
        (item: unknown): item is Observable<unknown> => !!item
      );
    return new Observable(sub => {
      sub.add(() => branch.teardown());

      if (observables.length < 1) {
        if (typeof index === 'number') {
          attacher.remove(index);
        }
        branch.activate();
        sub.next();
        return;
      }
      return combineLatest(observables)
        .pipe(
          tap(() => {
            if (typeof index === 'number') {
              attacher.remove(index);
            }
            branch.activate();
            sub.next();
          })
        )
        .subscribe();
    });
  } else if (isObservable(children)) {
    return children.pipe(
      switchMap(child => {
        const obs = attachChildren(attacher, child, index);
        if (!obs) {
          return of(void 0);
        }
        return obs;
      }),
      filter(onlyFirst)
    );
  }
}
