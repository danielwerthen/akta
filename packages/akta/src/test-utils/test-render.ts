import http from 'http';
import puppeteer, { Viewport } from 'puppeteer';

function getCSS(doc: Document) {
  const lines: string[] = [];
  const styleElements = doc.getElementsByTagName('style');
  for (let index = 0; index < styleElements.length; index++) {
    const sheet = styleElements.item(index)?.sheet;
    const media = styleElements.item(index)?.media;
    if (!sheet) {
      continue;
    }
    if (media) {
      lines.push(`@media ${media} {`);
    }
    for (let ruleIdx = 0; ruleIdx < (sheet.cssRules.length || 0); ruleIdx++) {
      const rule = sheet.cssRules[ruleIdx];
      if (rule?.cssText) {
        lines.push(rule?.cssText);
      }
    }
    if (media) {
      lines.push('}');
    }
  }

  return lines.join('\n');
}

export function createRenderer() {
  const port = 4352;
  return async function render(
    doc: Document,
    defaultViewport: Viewport = { width: 800, height: 600 }
  ) {
    const browser = await puppeteer.launch({
      defaultViewport,
    });
    const page = await browser.newPage();
    const server = http.createServer((_req, res) => {
      res.writeHead(200);
      res.end(`<html>
        <head>
          <link rel="stylesheet" href="https://unpkg.com/normalize.css@8.0.1/normalize.css" />
          <style>
            ${getCSS(doc)}
          </style>
        </head>
        ${doc.body.outerHTML}
      </html>`);
    });
    server.listen(port);
    try {
      await page.goto(`http://localhost:${port}`);
      return await page.screenshot();
    } finally {
      server.close();
      browser.close();
    }
  };
}
