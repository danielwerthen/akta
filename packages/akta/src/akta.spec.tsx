/** @jsxImportSource . */
import { mount } from './mount-ops';
import { toMatchImageSnapshot } from 'jest-image-snapshot';
import { createRenderer } from './test-utils/test-render';
import { useNext } from './dependencies';

function getCSS() {
  const lines: string[] = [];
  const styleElements = document.getElementsByTagName('style');
  for (let index = 0; index < styleElements.length; index++) {
    const sheet = styleElements.item(index)?.sheet;
    const media = styleElements.item(index)?.media;
    if (!sheet) {
      continue;
    }
    if (media) {
      lines.push(media);
    }
    for (let ruleIdx = 0; ruleIdx < (sheet.cssRules.length || 0); ruleIdx++) {
      const rule = sheet.cssRules[ruleIdx];
      if (rule?.cssText) {
        lines.push(rule?.cssText);
      }
    }
  }

  return lines.join('\n');
}

declare global {
  namespace jest {
    interface Matchers<R> {
      toMatchImageSnapshot(): R;
    }
  }
}

expect.extend({
  toMatchImageSnapshot,
});

describe('Akta', () => {
  let container: HTMLElement;
  const render = createRenderer();
  beforeEach(() => {
    container = document.createElement('div');
    container.id = 'app';
    document.body.append(container);
  });
  afterEach(() => {
    document.getElementById('app')?.remove();
  });
  function rit(str: string, func: () => (() => void) | Promise<() => void>) {
    it(str, async () => {
      const cleanup = await func();
      expect(document.body).toMatchSnapshot();
      expect(getCSS()).toMatchSnapshot();
      expect(
        await render(document, {
          width: 50,
          height: 50,
        })
      ).toMatchImageSnapshot({
        dumpDiffToConsole: true,
      });
      expect(
        await render(document, {
          width: 150,
          height: 150,
        })
      ).toMatchImageSnapshot();
      expect(
        await render(document, {
          width: 500,
          height: 500,
        })
      ).toMatchImageSnapshot();
      cleanup();
    });
  }
  rit('should handle css props with media queries properly', async () => {
    return mount(
      <div
        height="100vh"
        width="100vw"
        $backgroundColor="red"
        $backgroundColor_min100="blue"
        $backgroundColor_max500_min200="green"
      >
        <svg
          viewBox="0 0 50 50"
          xmlns="http://www.w3.org/2000/svg"
          stroke="red"
          fill="grey"
        >
          <circle cx="10" cy="10" r="5"></circle>
          <circle cx="20" cy="20" r="5"></circle>

          <svg
            viewBox="0 0 10 10"
            x="40"
            width="10"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle cx="5" cy="5" r="4"></circle>
          </svg>
        </svg>
      </div>,
      container
    );
  });

  it('it should render generator components', async () => {
    function* Component() {
      const next = useNext();
      yield (
        <div>
          <p>State 1</p>
          <button onClick={next}>Next</button>
        </div>
      );
      yield (
        <div>
          <p>State 2</p>
          <button onClick={next}>Next</button>
        </div>
      );
      return <p>State 3</p>;
    }
    const cleanup = mount(
      <div>
        <Component />
      </div>,
      container
    );

    document.body
      .getElementsByTagName('button')
      .item(0)
      ?.click();

    expect(document.body).toMatchSnapshot();

    document.body
      .getElementsByTagName('button')
      .item(0)
      ?.click();

    expect(document.body).toMatchSnapshot();
    cleanup();
  });
});
