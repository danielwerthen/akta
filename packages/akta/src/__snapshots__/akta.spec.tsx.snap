// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Akta it should render generator components 1`] = `
<body>
  <div
    id="app"
  >
    <div>
      <div>
        <p>
          State 2
        </p>
        <button>
          Next
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Akta it should render generator components 2`] = `
<body>
  <div
    id="app"
  >
    <div>
      <p>
        State 3
      </p>
    </div>
  </div>
</body>
`;

exports[`Akta should handle css props with media queries properly 1`] = `
<body>
  <div
    id="app"
  >
    <div
      class="c0 c1 c2"
      style="height: 100vh; width: 100vw;"
    >
      <svg
        style="stroke: red; fill: grey;"
        viewBox="0 0 50 50"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="10"
          cy="10"
          r="5"
        />
        <circle
          cx="20"
          cy="20"
          r="5"
        />
        <svg
          viewBox="0 0 10 10"
          x="40"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="5"
            cy="5"
            r="4"
          />
        </svg>
      </svg>
    </div>
  </div>
</body>
`;

exports[`Akta should handle css props with media queries properly 2`] = `
".c0 {background-color: red;}
(min-width: 100px)
.c1 {background-color: blue;}
(max-width: 500px) and (min-width: 200px)
.c2 {background-color: green;}"
`;
